__tests__/client/service-worker/events/utility.spec.js

```diff
@@ -0,0 +1,241 @@
+      revision: null,
+    }],
+  ])('extracts metadata from %s', (url, [type, baseUrl], result) => {
+    // eslint-disable-next-line no-param-reassign
+    if (result.revision === null) delete result.revision;
+    const meta = createResourceMetaData({ request: { url } }, [type, baseUrl], result.revision);
+    expect(meta).toEqual({ ...result, url });
+  });
+});
+
+describe(invalidateCacheResource.name, () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  test('gets the resource metadata and validates the incoming request', async () => {
+    const waitUntil = jest.fn();
+    const meta = { ...metaBluePrint };
+    const event = { request: { url: meta.url }, waitUntil };
+    const response = {};
+    const responseHandler = invalidateCacheResource(event, meta);
+
+    getMetaData.mockImplementationOnce(() => Promise.resolve({}));
+
+    expect(responseHandler(response)).toBe(response);
+    expect(waitUntil).toHaveBeenCalledTimes(1);
+    expect(getMetaData).toHaveBeenCalledTimes(1);
+    expect(getMetaData).toHaveBeenCalledWith({ cacheName: 'modules/assets/cat.jpg', url: 'http://localhost/modules/assets/cat.jpg' });
+    await waitUntil.mock.calls[0][0];
+    expect(setMetaData).toHaveBeenCalledTimes(1);
+    expect(remove).not.toHaveBeenCalled();
+  });
+
+  test('invalidates the incoming request due to version change', async () => {
+    const waitUntil = jest.fn();
+    const meta = { ...metaBluePrint };
+    const event = { request: { url: meta.url }, waitUntil };
+    const response = {};
+    const responseHandler = invalidateCacheResource(event, meta);
+
+    getMetaData.mockImplementationOnce(() => Promise.resolve({ ...meta, version: '0.0.5' }));
+
+    expect(responseHandler(response)).toBe(response);
+    expect(waitUntil).toHaveBeenCalledTimes(1);
+    expect(getMetaData).toHaveBeenCalledTimes(1);
+    expect(getMetaData).toHaveBeenCalledWith({ cacheName: 'modules/assets/cat.jpg', url: 'http://localhost/modules/assets/cat.jpg' });
+    await waitUntil.mock.calls[0][0];
+    expect(waitUntil).toHaveBeenCalledTimes(2);
+    expect(setMetaData).toHaveBeenCalledTimes(1);
+    expect(remove).toHaveBeenCalledTimes(1);
+  });
+});
+
+describe(setCacheResource.name, () => {
+  test('calls "put" on the cache with the resource', async () => {
+    const clone = jest.fn(() => 'clone');
+    const waitUntil = jest.fn();
+    const meta = { cacheName: metaBluePrint.cacheName };
+    const event = { request: { url: metaBluePrint.url, clone }, waitUntil };
+    const response = { clone };
+    const responseHandler = setCacheResource(event, meta);
+
+    expect(responseHandler(response)).toBe(response);
+    expect(waitUntil).toHaveBeenCalledTimes(1);
+    expect(clone).toHaveBeenCalledTimes(2);
+    expect(put).toHaveBeenCalledTimes(1);
+    expect(put).toHaveBeenCalledWith('clone', 'clone', meta);
+  });
+});
+
+describe(fetchCacheResource.name, () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
```

Comment:
these tests lean towards how the function is built more than what it's doing. 

`given match returns a response i expect to see the cached response and not call fetch.`

`given match does not return a cached response i expect fetch to have been called with X, set the result in the cache and ...`

Reply:
I'm not fully getting this point.. how would you test it?

Reply:
I made the `fetch.spec.js` to unit test the system/implementation as a whole, while `utility.spec.js` focuses on the portions of the system. Also, integration tests will cover the implementation.